# Eat Safe, Love

## Part 1: Database and Jupyter Notebook Set Up

Import the data provided in the `establishments.json` file from Terminal. Name the database `uk_food` and the collection `establishments`.

#### Importing establishments.json via Terminal and creating uk_food database
```
(base) crfitz@Colleens-Air nosql-challenge % cd Resources
(base) crfitz@Colleens-Air Resources % mongoimport --type json -d uk_food -c establishments --drop --jsonArray establishments.json
```

In [21]:
# Import dependencies
from pymongo import MongoClient
from pprint import pprint

In [22]:
# Create an instance of MongoClient
mongo = MongoClient(port=27017)

In [23]:
# confirm that our new database was created
mongo.list_database_names()

['admin',
 'classDB',
 'config',
 'gardenDB',
 'local',
 'met',
 'petsitly_marketing',
 'uk_food']

In [24]:
# assign the uk_food database to a variable name
db = mongo['uk_food']

In [25]:
# review the collections in our new database
db.list_collection_names()

['establishments']

In [26]:
# review a document in the establishments collection
db.establishments.find_one()

{'_id': ObjectId('648a6a149b0b7f46d1d87b87'),
 'FHRSID': 1034540,
 'ChangesByServerID': 0,
 'LocalAuthorityBusinessID': 'PI/000078691',
 'BusinessName': 'The Coastguard Inn',
 'BusinessType': 'Pub/bar/nightclub',
 'BusinessTypeID': 7843,
 'AddressLine1': 'The Bay',
 'AddressLine2': 'St Margarets Bay',
 'AddressLine3': 'Kent',
 'AddressLine4': '',
 'PostCode': 'CT15 6DY',
 'Phone': '',
 'RatingValue': '5',
 'RatingKey': 'fhrs_5_en-gb',
 'RatingDate': '2022-08-17T00:00:00',
 'LocalAuthorityCode': '182',
 'LocalAuthorityName': 'Dover',
 'LocalAuthorityWebSite': 'http://www.dover.gov.uk/',
 'LocalAuthorityEmailAddress': 'publicprotection@dover.gov.uk',
 'scores': {'Hygiene': 0, 'Structural': 5, 'ConfidenceInManagement': 0},
 'SchemeType': 'FHRS',
 'geocode': {'longitude': '1.387974', 'latitude': '51.152225'},
 'RightToReply': '',
 'Distance': 4587.347174863443,
 'NewRatingPending': False,
 'meta': {'dataSource': None,
  'extractDate': '0001-01-01T00:00:00',
  'itemCount': 0,
  'returncode'

In [27]:
# assign the collection to a variable
establishments = db['establishments']


## Part 2: Update the Database

1. An exciting new halal restaurant just opened in Greenwich, but hasn't been rated yet. The magazine has asked you to include it in your analysis. Add the following restaurant "Penang Flavours" to the database.

1A. Create a dictionary for the new restaurant data
(Changing this cell to Markdown to avoid creating duplicate entries - learned from experience)
new_restaurant = {"BusinessName":"Penang Flavours",
    "BusinessType":"Restaurant/Cafe/Canteen",
    "BusinessTypeID":"",
    "AddressLine1":"Penang Flavours",
    "AddressLine2":"146A Plumstead Rd",
    "AddressLine3":"London",
    "AddressLine4":"",
    "PostCode":"SE18 7DY",
    "Phone":"",
    "LocalAuthorityCode":"511",
    "LocalAuthorityName":"Greenwich",
    "LocalAuthorityWebSite":"http://www.royalgreenwich.gov.uk",
    "LocalAuthorityEmailAddress":"health@royalgreenwich.gov.uk",
    "scores":{
        "Hygiene":"",
        "Structural":"",
        "ConfidenceInManagement":""
    },
    "SchemeType":"FHRS",
    "geocode":{
        "longitude":"0.08384000",
        "latitude":"51.49014200"
    },
    "RightToReply":"",
    "Distance":4623.9723280747176,
    "NewRatingPending":True}

1B. Inserting new dictionary (changing to Markdown for same reason as previous cell)
db.establishments.insert_one(new_restaurant)

In [30]:
# Check that the new restaurant was inserted
query = {"BusinessName":"Penang Flavours"}
results = establishments.find(query)
for result in results:
    print(result)

{'_id': ObjectId('648a6a9280061f32712e9e31'), 'BusinessName': 'Penang Flavours', 'BusinessType': 'Restaurant/Cafe/Canteen', 'BusinessTypeID': '', 'AddressLine1': 'Penang Flavours', 'AddressLine2': '146A Plumstead Rd', 'AddressLine3': 'London', 'AddressLine4': '', 'PostCode': 'SE18 7DY', 'Phone': '', 'LocalAuthorityCode': '511', 'LocalAuthorityName': 'Greenwich', 'LocalAuthorityWebSite': 'http://www.royalgreenwich.gov.uk', 'LocalAuthorityEmailAddress': 'health@royalgreenwich.gov.uk', 'scores': {'Hygiene': '', 'Structural': '', 'ConfidenceInManagement': ''}, 'SchemeType': 'FHRS', 'geocode': {'longitude': '0.08384000', 'latitude': '51.49014200'}, 'RightToReply': '', 'Distance': 4623.972328074718, 'NewRatingPending': True}


2. Find the BusinessTypeID for "Restaurant/Cafe/Canteen" and return only the `BusinessTypeID` and `BusinessType` fields.

In [31]:
# Find the BusinessTypeID for "Restaurant/Cafe/Canteen" and return only the BusinessTypeID and BusinessType fields 
query = {"BusinessType":"Restaurant/Cafe/Canteen"}
fields = {'BusinessTypeID': 1, 'BusinessType': 1}
results = establishments.find(query, fields)
pprint(results.next())


{'BusinessType': 'Restaurant/Cafe/Canteen',
 'BusinessTypeID': 1,
 '_id': ObjectId('648a6a149b0b7f46d1d87b88')}


3. Update the new restaurant with the `BusinessTypeID` you found.

In [32]:
# Update the new restaurant with the correct BusinessTypeID
establishments.update_one({"BusinessName" : "Penang Flavours"},{"$set": {"BusinessTypeID":1}})


In [33]:
# Confirm that the new restaurant was updated
query = {"BusinessName":"Penang Flavours"}
results = establishments.find_one(query)
pprint(results)


{'AddressLine1': 'Penang Flavours',
 'AddressLine2': '146A Plumstead Rd',
 'AddressLine3': 'London',
 'AddressLine4': '',
 'BusinessName': 'Penang Flavours',
 'BusinessType': 'Restaurant/Cafe/Canteen',
 'BusinessTypeID': 1,
 'Distance': 4623.972328074718,
 'LocalAuthorityCode': '511',
 'LocalAuthorityEmailAddress': 'health@royalgreenwich.gov.uk',
 'LocalAuthorityName': 'Greenwich',
 'LocalAuthorityWebSite': 'http://www.royalgreenwich.gov.uk',
 'NewRatingPending': True,
 'Phone': '',
 'PostCode': 'SE18 7DY',
 'RightToReply': '',
 'SchemeType': 'FHRS',
 '_id': ObjectId('648a6a9280061f32712e9e31'),
 'geocode': {'latitude': '51.49014200', 'longitude': '0.08384000'},
 'scores': {'ConfidenceInManagement': '', 'Hygiene': '', 'Structural': ''}}


4. The magazine is not interested in any establishments in Dover, so check how many documents contain the Dover Local Authority. Then, remove any establishments within the Dover Local Authority from the database, and check the number of documents to ensure they were deleted.

In [34]:
# 4A. Find how many documents have LocalAuthorityName as "Dover"
dover_count = establishments.count_documents({'LocalAuthorityName': 'Dover'})
print(f"Number of Documents with LocalAuthorityName as Dover:{dover_count}")

Number of Documents with LocalAuthorityName as Dover:994


4A. Number of Documents with LocalAuthorityName as Dover before step 4B: 994

In [36]:
#4Ai. Find how many total documents exist in establishments to check later steps
total_doc_count = establishments.count_documents({})
print(f"Total Number of Documents: {total_doc_count}")


Total Number of Documents: 39780


4Ai. Total Number of Documents before deleting Dover: 39780


In [37]:
# 4B. Delete all documents where LocalAuthorityName is "Dover"
establishments.delete_many({'LocalAuthorityName': 'Dover'})

In [38]:
# 4Bi. Check if any remaining documents include Dover
query=db.establishments.find_one({'LocalAuthorityName': 'Dover'})
pprint (query)

None


In [39]:
#4Bii. Find how many total documents exist in establishments after deletion of Dover
total_doc_count = establishments.count_documents({})
print(f"Total Number of Documents: {total_doc_count}")


Total Number of Documents: 38786


In [40]:
# Check that other documents remain with 'find_one'
db.establishments.find_one()

{'_id': ObjectId('648a6a149b0b7f46d1d87e6c'),
 'FHRSID': 647177,
 'ChangesByServerID': 0,
 'LocalAuthorityBusinessID': 'PI/000041489',
 'BusinessName': 'Wear Bay Bowls Club',
 'BusinessType': 'Pub/bar/nightclub',
 'BusinessTypeID': 7843,
 'AddressLine1': 'Wear Bay Road',
 'AddressLine2': 'Folkestone',
 'AddressLine3': 'Kent',
 'AddressLine4': '',
 'PostCode': 'CT19 6PY',
 'Phone': '',
 'RatingValue': '4',
 'RatingKey': 'fhrs_4_en-gb',
 'RatingDate': '2014-03-31T00:00:00',
 'LocalAuthorityCode': '188',
 'LocalAuthorityName': 'Folkestone and Hythe',
 'LocalAuthorityWebSite': 'http://www.folkestone-hythe.gov.uk',
 'LocalAuthorityEmailAddress': 'foodteam@folkestone-hythe.gov.uk',
 'scores': {'Hygiene': 5, 'Structural': 5, 'ConfidenceInManagement': 10},
 'SchemeType': 'FHRS',
 'geocode': {'longitude': '1.196408', 'latitude': '51.086058'},
 'RightToReply': '',
 'Distance': 4591.821311183521,
 'NewRatingPending': False,
 'meta': {'dataSource': None,
  'extractDate': '0001-01-01T00:00:00',
  '

5. Some of the number values are stored as strings, when they should be stored as numbers.

Use `update_many` to convert `latitude` and `longitude` to decimal numbers.

In [41]:
# Change the data type from String to Decimal for longitude and latitude
establishments.update_many({}, [{'$set':{ "geocode.latitude" : {'$toDouble': "$geocode.latitude"}} }])
establishments.update_many({}, [{'$set':{ "geocode.longitude" : {'$toDouble': "$geocode.longitude"}} }])



Use `update_many` to convert `RatingValue` to integer numbers.

In [42]:
# Set non 1-5 Rating Values to Null
non_ratings = ["AwaitingInspection", "Awaiting Inspection", "AwaitingPublication", "Pass", "Exempt"]
establishments.update_many({"RatingValue": {"$in": non_ratings}}, [ {'$set':{ "RatingValue" : None}} ])


In [43]:
# Change the data type from String to Integer for RatingValue
establishments.update_many({}, [{'$set':{ "RatingValue" : {'$toInt': "$RatingValue"}} }])


In [44]:
# Check that the coordinates and rating value are now numbers
db.establishments.find_one()

{'_id': ObjectId('648a6a149b0b7f46d1d87e6c'),
 'FHRSID': 647177,
 'ChangesByServerID': 0,
 'LocalAuthorityBusinessID': 'PI/000041489',
 'BusinessName': 'Wear Bay Bowls Club',
 'BusinessType': 'Pub/bar/nightclub',
 'BusinessTypeID': 7843,
 'AddressLine1': 'Wear Bay Road',
 'AddressLine2': 'Folkestone',
 'AddressLine3': 'Kent',
 'AddressLine4': '',
 'PostCode': 'CT19 6PY',
 'Phone': '',
 'RatingValue': 4,
 'RatingKey': 'fhrs_4_en-gb',
 'RatingDate': '2014-03-31T00:00:00',
 'LocalAuthorityCode': '188',
 'LocalAuthorityName': 'Folkestone and Hythe',
 'LocalAuthorityWebSite': 'http://www.folkestone-hythe.gov.uk',
 'LocalAuthorityEmailAddress': 'foodteam@folkestone-hythe.gov.uk',
 'scores': {'Hygiene': 5, 'Structural': 5, 'ConfidenceInManagement': 10},
 'SchemeType': 'FHRS',
 'geocode': {'longitude': 1.196408, 'latitude': 51.086058},
 'RightToReply': '',
 'Distance': 4591.821311183521,
 'NewRatingPending': False,
 'meta': {'dataSource': None,
  'extractDate': '0001-01-01T00:00:00',
  'itemCo